tests/test_scripts.py

```diff
@@ -0,0 +1,372 @@
+        b"fatal: Remote branch nonexistent not found in upstream origin\n"
+        == e.value.stderr
+    )
+
+
+def test_non_qubesos_repository(temp_directory):
+    args = create_dummy_args(
+        component_directory=temp_directory,
+        component_repository="https://github.com/fepitre/qubes-core-qrexec",
+        git_branch="builderv2",
+    )
+    with pytest.raises(ValueError) as e:
+        get_and_verify_source(args)
+    (msg,) = e.value.args
+    assert msg == "---> Invalid commit 961bf3c7bd8cc5dc5335ee91afefbf7feb92c982."
+
+
+def test_non_qubesos_repository_with_maintainer_and_no_signed_tags(
+    temp_directory,
+):
+    args = create_dummy_args(
+        component_directory=temp_directory,
+        component_repository="https://github.com/fepitre/qubes-core-qrexec",
+        git_branch="builderv2",
+        less_secure_signed_commits_sufficient=True,
+        maintainers=["9FA64B92F95E706BF28E2CA6484010B5CDC576E2"],
+    )
+    get_and_verify_source(args)
+    assert (
+        subprocess.run(
+            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
+            capture_output=True,
+            cwd=str(temp_directory),
+            text=True,
+        ).stdout.strip()
+        == "builderv2"
+    )
+
+
+def test_repository_no_version_tags(temp_directory):
+    args = create_dummy_args(
+        component_repository="https://github.com/fepitre/qubes-core-qrexec",
+        component_directory=temp_directory,
+        git_branch="builderv2-tests-1",
+        fetch_versions_only=True,
+    )
+    with pytest.raises(subprocess.CalledProcessError) as e:
+        get_and_verify_source(args)
+    assert "fatal: No names found, cannot describe anything.\n" == e.value.stderr
+
+
+def test_existing_repository_no_version_tags(temp_directory):
+    component_repository = "https://github.com/fepitre/qubes-core-qrexec"
+    git_branch = "builderv2-tests-1"
+    subprocess.run(
+        ["git", "clone", "-b", git_branch, component_repository, temp_directory],
+        check=True,
+        capture_output=True,
+    )
+    args = create_dummy_args(
+        component_repository=component_repository,
+        component_directory=temp_directory,
+        git_branch=git_branch,
+        fetch_versions_only=True,
+    )
+    get_and_verify_source(args)
+    assert not (temp_directory / ".git/FETCH_HEAD").exists()
+
+
+def test_existing_repository_with_version_tags(temp_directory):
+    component_repository = "https://github.com/fepitre/qubes-core-qrexec"
+    git_branch = "builderv2-tests-2"
+    subprocess.run(
+        ["git", "clone", "-b", git_branch, component_repository, temp_directory],
+        check=True,
+        capture_output=True,
+    )
+    args = create_dummy_args(
+        component_repository=component_repository,
+        component_directory=temp_directory,
+        git_branch=git_branch,
+        fetch_versions_only=True,
+    )
+    get_and_verify_source(args)
+    assert (temp_directory / ".git/FETCH_HEAD").exists()
```

Comment:
Check also if really version tag is on top. You might want to adjust initial clone (or add `git reset` just after) to have it few commits before version tag.

Similarly, a useful test would be with fresh clone with fetch_versions_only (pointing at branch where version tag is not on top but sometime earlier - and check if really that version was fetched).